src/pkg/cli/subscribe_test.go

```diff
@@ -0,0 +1,113 @@
+package cli
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/DefangLabs/defang/src/pkg"
+	"github.com/DefangLabs/defang/src/pkg/cli/client"
+	defangv1 "github.com/DefangLabs/defang/src/protos/io/defang/v1"
+)
+
+// MockSubscribeServerStream mocks the stream response for Subscribe.
+type MockSubscribeServerStream struct {
+	index int
+	Resps []*defangv1.SubscribeResponse
+	Error error
+}
+
+func (*MockSubscribeServerStream) Close() error {
+	return nil
+}
+
+func (m *MockSubscribeServerStream) Receive() bool {
+	if m.index >= len(m.Resps) {
+		return false
+	}
+	m.index++
+	return true
+}
+
+func (m *MockSubscribeServerStream) Msg() *defangv1.SubscribeResponse {
+	if m.index == 0 || m.index > len(m.Resps) {
+		return nil
+	}
+	return m.Resps[m.index-1]
+}
+
+func (m *MockSubscribeServerStream) Err() error {
+	return m.Error
+}
+
+// mockSubscribeProvider mocks the provider for Subscribe.
+type mockSubscribeProvider struct {
+	client.MockProvider
+	Reqs []*defangv1.SubscribeRequest
+}
+
+func (m *mockSubscribeProvider) Subscribe(
+	_ context.Context,
+	req *defangv1.SubscribeRequest,
+) (client.ServerStream[defangv1.SubscribeResponse], error) {
+	m.Reqs = append(m.Reqs, req)
+
+	stream := &MockSubscribeServerStream{
+		Resps: []*defangv1.SubscribeResponse{
+			{Name: "service1", State: defangv1.ServiceState_DEPLOYMENT_COMPLETED},
+			{Name: "service2", State: defangv1.ServiceState_BUILD_FAILED},
```

Comment:
It would be nice if we can also cover the `defangv1.ServiceState_DEPLOYMENT_FAILED` state

Reply:
minor: it would also be nice to cover the case of other service states other than `COMPLETED` or `FAILED` to check only all completed or single failed would terminate the wait.